Add `.to_best_block()` method to `ValidatedBlockHeader`

MaxFangX opened the pull request:
Adds a new method `.to_best_block()` to `ValidatedBlockHeader` which returns a `BestBlock` that nodes can pass into the channel manager during a fresh start, in order to prevent a possible race condition which eventually results in a channel manager panic when a node is started for the first time.

Required for the bugfix in https://github.com/lightningdevkit/ldk-sample/pull/78.

jkczyz commented on lightning-block-sync/src/init.rs:
```diff
@@ -26,6 +26,47 @@ BlockSourceResult<ValidatedBlockHeader> where B::Target: BlockSource {
+	let (polled_best_block_hash, maybe_best_block_height) = block_source
+		.get_best_block()
+		.await?;
+	let best_block_header_data = block_source
+		.get_header(&polled_best_block_hash, maybe_best_block_height)
+		.await?;
+	let best_block_height = maybe_best_block_height.ok_or_else(||
+			BlockSourceError::persistent("block height missing")
+		)?;
```
`BlockHeaderData` includes the height, so no need to fail. This height is just a hint for the `BlockSource` when fetching the header.

MaxFangX replied:
Removed this function entirely as per your suggestion

jkczyz commented on lightning-block-sync/src/init.rs:
```diff
@@ -26,6 +26,47 @@ BlockSourceResult<ValidatedBlockHeader> where B::Target: BlockSource {
+// to hear about the intermediate block, thereby panicking with the message
+// "Blocks must be connected in chain-order - the connected header must build on
+// the last connected header."
+pub async fn validate_best_block_header_and_best_block<B: Deref>(block_source: B)
```
To avoid adding another function to `init`, we could have a conversion method on `ValidatedBlockHeader` that returns a `BestBlock`.

MaxFangX replied:
Ah nice, that's a much cleaner way to do this.

MaxFangX commented:
Pushed changes @jkczyz, removed the new `init` function entirely and replaced it with a new `.best_block()` method on `ValidatedBlockHeader`. It's a lot cleaner this way, thanks for the suggestion!

jkczyz commented on lightning-block-sync/src/poll.rs:
```diff
@@ -140,6 +141,19 @@ impl ValidatedBlockHeader {
+    /// [`ChainParameters`]: lightning::ln::channelmanager::ChainParameters
+    /// [`ChannelManager`]: lightning::ln::channelmanager::ChannelManager
+    /// [`ChannelManager::new`]: lightning::ln::channelmanager::ChannelManager::new
+    pub fn best_block(&self) -> BestBlock {
```
Let's name this `to_best_block` since it is a borrowed-to-owned conversion, rather than a property accessor.

https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

jkczyz commented on lightning-block-sync/src/poll.rs:
```diff
@@ -140,6 +141,19 @@ impl ValidatedBlockHeader {
 		Ok(())
 	}
+
+    /// Returns the [`BestBlock`] corresponding to this validated block header,
+    /// which can be passed into [`ChannelManager::new`] as part of its
+    /// [`ChainParameters`]. Useful for ensuring that the [`SpvClient`] and
+    /// [`ChannelManager`] are synced to the same block during a fresh start.
```
Could you wrap at 100 characters? (I know rustfmt would have fixed this 😛 )

jkczyz replied:
nit: s/synced/initialized

(Technically, `ChannelManager` is being synced by `SpvClient`.)

MaxFangX commented:
Implemented once again @jkczyz

jkczyz reviewed (approved):
Thanks for putting this together!

tnull reviewed (approved):
LGTM
